cmd/jujud/updateseries/updateseries.go

```diff
@@ -0,0 +1,396 @@
+}
+
+func (c *UpdateSeriesCommand) Run(ctx *cmd.Context) error {
+	if err := c.findAgents(); err != nil {
+		return err
+	}
+
+	fromInitSys, err := service.VersionInitSystem(c.fromSeries)
+	if err != nil {
+		return err
+	}
+
+	toInitSys, err := service.VersionInitSystem(c.toSeries)
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case toInitSys == service.InitSystemUpstart:
+		return errors.NewNotSupported(nil, "downgrade to series using upstart not supported")
+	case toInitSys == fromInitSys && toInitSys == service.InitSystemSystemd:
+		return errors.Errorf("no work to be done, both series use systemd")
+	case toInitSys == service.InitSystemSystemd:
+		errorHappened := false
+		if err = c.writeSystemdAgents(ctx); err != nil {
+			ctx.Warningf("%s", err)
+			errorHappened = true
+		}
+		if err = c.copyTools(); err != nil {
+			ctx.Warningf("%s", err)
+			errorHappened = true
+		} else {
+			ctx.Infof("successfully copied tools and relinked agent tools")
+		}
+		switch {
+		case !errorHappened && c.startAgents:
+			if err = c.startAllAgents(ctx); err != nil {
+				return err
+			}
+			ctx.Infof("all agents successfully restarted")
+		case errorHappened && c.startAgents:
+			return errors.Errorf("anable to start agents due to previous errors")
+		}
+	default:
+		return errors.Errorf("unknown happened")
+	}
+	return nil
+}
+
+func (c *UpdateSeriesCommand) findAgents() error {
+	agentsDir := filepath.Join(c.dataDir, "agents")
+	dir, err := os.Open(agentsDir)
+	if err != nil {
+		return errors.Annotate(err, "opening agents dir")
+	}
+	defer dir.Close()
+
+	entries, err := dir.Readdir(-1)
+	if err != nil {
+		return errors.Annotate(err, "reading agents dir")
+	}
+	for _, info := range entries {
+		name := info.Name()
+		tag, err := names.ParseTag(name)
+		if err != nil {
+			continue
+		}
+		if tag.Kind() == names.MachineTagKind {
+			c.machineAgent = name
+		} else if tag.Kind() == names.UnitTagKind {
+			c.unitAgents = append(c.unitAgents, name)
+		}
+	}
+	return nil
+}
+
+var isController = func() (bool, error) {
+	services, err := service.ListServices()
+	if err != nil {
+		return true, err
+	}
+	for _, service := range services {
+		if strings.HasPrefix(service, mongo.ServiceName) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+var (
+	systemdDir          = "/etc/systemd/system"
+	systemdMultiUserDir = systemdDir + "/multi-user.target.wants"
+)
+
+// For testing
+var sysdIsRunning = systemd.IsRunning
+
+func (c *UpdateSeriesCommand) writeSystemdAgents(ctx *cmd.Context) error {
+	var lastError error
+	for _, agentName := range append(c.unitAgents, c.machineAgent) {
+		conf, err := c.createAgentConf(agentName)
+		if err != nil {
+			ctx.Warningf("%s", err)
+			lastError = err
+			continue
+		}
+		svcName := "jujud-" + agentName
+		svc, err := service.NewService(svcName, conf, c.toSeries)
+
+		upgradableSvc, ok := svc.(service.UpgradableService)
+		if !ok {
+			return errors.Errorf("service not of type UpgradableService")
```

Comment:
Let's include svcName in this message for clarity.

Reply:
svcName isn't useful here because you can have a jujud-machine-0 service that is upgradable and one that isn't too.... upgradable is based on the host series.  I've updated the error to include upstart/systemd/windows as appropriate if this is hit.